File: 0002 Add Two Numbers/AddTwoNumbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil {
		if l2 == nil {
			return nil
		} else {
			return l2
		}
	} else if l2 == nil {
		return l1
	}

	ret := l1
	sum := l1.Val + l2.Val
	l1.Val = sum % 10
	tens := sum / 10

	for l1.Next != nil && l2.Next != nil {
		l1 = l1.Next
		l2 = l2.Next

		sum = l1.Val + l2.Val + tens
		l1.Val = sum % 10
		tens = sum / 10
	}

	if l1.Next == nil {
		if l2.Next == nil {
			if tens != 0 {
				l1.Next = new(ListNode)
				l1.Next.Val = tens
				return ret
			}
			return ret
		}
		l1.Next = l2.Next
	}

	for l1.Next != nil {
		l1 = l1.Next

		sum = l1.Val + tens
		l1.Val = sum % 10
		tens = sum / 10
	}

	if tens != 0 {
		l1.Next = new(ListNode)
		l1.Next.Val = tens
	}
	return ret
}

File: 0002 Add Two Numbers/AddTwoNumbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    if (!l1)
        return l2 ? NULL : l2;
    else if (!l2)
        return l1;
    
    struct ListNode *ret = l1;
    int sum = l1->val + l2->val;
    l1->val = sum % 10;
    int tens = sum / 10;

    while (l1->next && l2->next) {
        l1 = l1->next;
        l2 = l2->next;

        sum = l1->val + l2->val + tens;
        l1->val = sum % 10;
        tens = sum / 10;
    }

    if (!l1->next) {
        if (!l2->next) {
            if (tens != 0) {
                l1->next = (struct ListNode*)malloc(sizeof(struct ListNode));
                l1 = l1->next
                l1->val = tens;
                l1->next = NULL;
                return ret;
            }
            return ret;
        }
        l1->next = l2->next;
    }

    do {
        l1 = l1->next;

        sum = l1->val + tens;
        l1->val = sum % 10;
        tens = sum / 10;

    } while (l1->next);
    
    if (tens != 0) {
        l1->next = (struct ListNode*)malloc(sizeof(struct ListNode));
        l1 = l1->next
        l1->val = tens;
        l1->next = NULL;
    }
    return ret;
}